xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -340,111 +328,90 @@ public boolean isEquivalentTo(RoundRobinPicker picker) {
-   *       A=10.
-   *   <li>Remaining deadlines are A=10 and B=10, so A is returned. The deadline of A is updated
-   *      with A=12.
-   *   <li>Remaining deadlines are A=12 and B=10, so B is returned. The deadline of B is updated
-   *      with B=15.
-   *   <li>etc.
-   * </ul>
-   *
-   * <p>In short: the entry with the highest weight is preferred.
-   *
-   * <ul>
-   *   <li>add() - O(lg n)
-   *   <li>pick() - O(lg n)
-   * </ul>
-   *
+   *  <li>nextSequence() - O(1)
+   *  <li>pick() - O(n)
    */
   @VisibleForTesting
-  static final class EdfScheduler {
-    private final PriorityQueue<ObjectState> prioQueue;
-
-    /**
-     * Weights below this value will be upped to this minimum weight.
-     */
-    private static final double MINIMUM_WEIGHT = 0.0001;
-
-    private final Object lock = new Object();
-
+  static final class StaticStrideScheduler {
+    private final int[] scaledWeights;
+    private final int sizeDivisor;
     private final Random random;
+    private final AtomicInteger sequence;
+    private static final int K_MAX_WEIGHT = 0xFFFF;
+    private static final long UINT32_MAX = 0xFFFF_FFFFL;
+
+    StaticStrideScheduler(float[] weights, Random random) {
+      int numChannels = weights.length;
+      checkArgument(numChannels >= 1, "Couldn't build scheduler: requires at least one weight");
+      int numWeightedChannels = 0;
+      double sumWeight = 0;
+      float maxWeight = 0;
+      int meanWeight = 0;
+      for (float weight : weights) {
+        if (weight > 0.0001) {
+          sumWeight += weight;
+          maxWeight = Math.max(weight, maxWeight);  
+          numWeightedChannels++;
+        }
+      }
 
-    /**
-     * Use the item's deadline as the order in the priority queue. If the deadlines are the same,
-     * use the index. Index should be unique.
-     */
-    EdfScheduler(int initialCapacity, Random random) {
-      this.prioQueue = new PriorityQueue<ObjectState>(initialCapacity, (o1, o2) -> {
-        if (o1.deadline == o2.deadline) {
-          return Integer.compare(o1.index, o2.index);
+      double scalingFactor = K_MAX_WEIGHT / maxWeight;
+      if (numWeightedChannels > 0) {
+        meanWeight = (int) Math.round(scalingFactor * sumWeight / numWeightedChannels);
+      } else {
+        meanWeight = 1;
+      }
+
+      // scales weights s.t. max(weights) == K_MAX_WEIGHT, meanWeight is scaled accordingly
+      int[] scaledWeights = new int[numChannels];
+      for (int i = 0; i < numChannels; i++) {
+        if (weights[i] < 0.0001) {
+          scaledWeights[i] = meanWeight;
         } else {
-          return Double.compare(o1.deadline, o2.deadline);
+          scaledWeights[i] = (int) Math.round(weights[i] * scalingFactor);
         }
-      });
+      }
+
+      this.scaledWeights = scaledWeights;
+      this.sizeDivisor = numChannels;
       this.random = random;
+      this.sequence = new AtomicInteger((int) (this.random.nextDouble() * UINT32_MAX));
```

Comment:
I just noticed that this may generate negative value when the high bit is 1, e,g. `new AtomicInteger((int)2147483649L)` is `-2147483647`.

Reply:
So I haven't pushed the change yet because I'm still testing, but I changed this to `this.sequence = new AtomicInteger(this.random.nextInt());`. 

The previous code worked strangely due to the interaction of a double with a long as well as the int cast from a high value, and didn't achieve the desired behavior of generating a random number between 0 and `UINT32_MAX`.

The new code gives us values ranging from `Integer.MIN_VALUE` to `Integer.MAX_VALUE`. I believe this is the desired behavior, because when the sequence value is casted up in `nextSequence()`, negative values become positive values that become unsigned integers greater than `Integer.MAX_VALUE` due to the `Integer.toUnsignedLong()`.

Reply:
It sounds fine. (You can just use `random` instead of `this.random`)